Port top navigation bar to Bootstrap

matjack1 opened the pull request:
I've tried to keep this PR as small as possible to start the new Bootstrap UI project. The idea is to swap the navigation with this version that is responsive. Is not adding anything yet to pages that are not responsive, but now we can start to port one page after another.

A couple of comments regarding the code:
- I've introduced font-awesome inspired by the new-gui branch, but it's not essential. I think it's better looking and better to use than what we had before, but it's just my opinion.
- in the bootrstrap_and_overrides CSS I'm now just making bootstrap work with the current CSS. Once the transition is complete we could delete this file.
- I had to rename the badge class because it conflicted with a badge class in Bootstrap.
- I've not ported the colours from the old navigation, but I think it would be a good idea, to preserve a bit the style. I'm opening the PR to have a conversation on the code, but it's still a WIP.
- if you could try on your computers and mobiles would be great!

One thing that I think it's good on this new version is that it's a bit smaller than the previous nav, saving 20px at the top, which is nice.

dnrce commented on app/assets/stylesheets/bootstrap_and_overrides.css.scss:
```diff
@@ -0,0 +1,13 @@
+.bootstrap {
```
I don't care for any of these global overrides. I think we'll be best off if we stay close to vanilla Bootstrap until we're 100% migrated.

If they were to remedy conflicts with Tracks styles, check out #1924.

matjack1 replied:
Yes, they were just remedies to conflicts, I'll get rid of everything thanks to your PR.

dnrce commented on app/assets/stylesheets/print.css.scss:
```diff
@@ -16,7 +16,7 @@ img {
 }
 
-#navcontainer, #input_box, #footer, .big-box, .refresh, .badge, h1, .icon,
+#navcontainer, #input_box, #footer, .big-box, .refresh, .number_badge, h1, .icon,
```
Hyphens seem to be more conventional than underscores in CSS class names.

dnrce replied:
...though this change should be unnecessary with #1924.

dnrce commented on app/views/shared/_navbar.html.erb:
```diff
@@ -0,0 +1,80 @@
+<nav id="navcontainer" class="navbar navbar-default navbar-fixed-top">
```
`navbar-inverse` instead of `navbar-default` will give a dark background more in line with the existing Tracks look & feel.

dnrce replied:
Drop the legacy `#navcontainer`.

dnrce commented on Gemfile.lock:
```diff
@@ -242,6 +244,7 @@ DEPENDENCIES
   cucumber-rails
   database_cleaner
   factory_girl_rails
+  font-awesome-sass (~> 4.4.0)
```
My vote is to avoid adding Font Awesome for now. Let's limit the scope of this PR and replace icons in a later changeset.

dnrce commented on Gemfile:
```diff
@@ -44,6 +44,7 @@ gem 'bcrypt', '~> 3.1.7'
 # gem 'jbuilder', '~> 1.2'
 
 gem "bootstrap-sass", "3.3.3"
+gem "font-awesome-sass", "~> 4.4.0"
```
My vote is to avoid adding Font Awesome for now. Let's limit the scope of this PR and replace icons in a later changeset.

dnrce commented:
Great start!

I think #1924 will turn this into a really clean diff by removing the need to add overrides and change `.badge`.

dnrce commented on app/views/shared/_navbar.html.erb:
```diff
@@ -0,0 +1,80 @@
+        <li><%= navigation_link(icon('search'), search_path, :title => t('layouts.navigation.search')) %></li>
+      </ul>
+      <ul class="nav navbar-nav navbar-right">
+        <li id="minilinks">
```
I think you can drop `#minilinks`.

dnrce commented on app/views/shared/_navbar.html.erb:
```diff
@@ -0,0 +1,80 @@
+<nav id="navcontainer" class="navbar navbar-default navbar-fixed-top">
+  <div class="container-fluid">
+    <div class="navbar-header">
+      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
```
I don't *think* the `aria-expanded` properties are needed because this defaults to `false`.

dnrce replied:
Eh, they're in the Bootstrap examples, so stick with this as-is.

dnrce commented on app/views/shared/_navbar.html.erb:
```diff
@@ -0,0 +1,80 @@
+            <li><%= navigation_link( t('layouts.navigation.recurring_todos'), {:controller => "recurring_todos", :action => "index"}, :title => t('layouts.navigation.recurring_todos_title')) %></li>
+          </ul>
+        </li>
+        <li id="menu_view" class="dropdown">
```
`#menu_view` is legacy and can be removed.

dnrce commented on app/views/shared/_navbar.html.erb:
```diff
@@ -0,0 +1,80 @@
+      </ul>
+      <ul class="nav navbar-nav navbar-right">
+        <li id="minilinks">
+          <%= link_to("#{t('common.logout')} (#{current_user.display_name}) &raquo;".html_safe, logout_path, :class => 'navbar-link') %>
```
Is the `.navbar-link` class needed? I don't see that it actually applies any styles.

dnrce commented on app/views/shared/_navbar.html.erb:
```diff
@@ -0,0 +1,80 @@
+            <li><%= link_to t('layouts.navigation.api_docs'), rest_api_docs_path %></li>
+          </ul>
+        </li>
+        <li><%= navigation_link(icon('search'), search_path, :title => t('layouts.navigation.search')) %></li>
```
Leave this as `image_tag` if not adding Font Awesome yet.

dnrce commented on app/views/shared/_navbar.html.erb:
```diff
@@ -0,0 +1,80 @@
+            <li><%= navigation_link( t('layouts.navigation.completed_tasks'), done_overview_path, {:accesskey=>"d", :title=>t('layouts.navigation.completed_tasks_title')} ) %></li>
+            <li><%= navigation_link( t('layouts.navigation.feeds'), feeds_path, :title => t('layouts.navigation.feeds_title')) %></li>
+            <li><%= navigation_link( t('layouts.navigation.stats'), stats_path, :title => t('layouts.navigation.stats_title')) %></li>
+            <li><hr/></li>
```
You can use Bootstrap's `<li class="divider"></li>` here.

dnrce commented on app/views/shared/_navbar.html.erb:
```diff
@@ -0,0 +1,80 @@
+            <li><%= navigation_link( t('layouts.navigation.feeds'), feeds_path, :title => t('layouts.navigation.feeds_title')) %></li>
+            <li><%= navigation_link( t('layouts.navigation.stats'), stats_path, :title => t('layouts.navigation.stats_title')) %></li>
+            <li><hr/></li>
+            <li id="menu_view_toggle_contexts"><%= link_to(t('layouts.toggle_contexts'), "#", {:title => t('layouts.toggle_contexts_title'), :id => "toggle-contexts-nav"}) %></li>
```
`#menu_view_toggle_contexts` is legacy.

dnrce commented on app/views/shared/_navbar.html.erb:
```diff
@@ -0,0 +1,80 @@
+          </ul>
+        </li>
+        <li id="menu_view" class="dropdown">
+          <a href="#" id="menu_view_link" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">
```
`#menu_view_link` is legacy.

dnrce commented on app/views/shared/_navbar.html.erb:
```diff
@@ -0,0 +1,80 @@
+        <%= l(Time.zone.today, :format => current_user.prefs.title_date_format) %>
+      </div>
+    </div>
+    <div id="navbar" class="navbar-collapse collapse" aria-expanded="false">
```
`aria-expanded` is not, however, needed on the target element. In the Bootstrap docs it's only placed on the controlling element.

dnrce commented on app/views/shared/_navbar.html.erb:
```diff
@@ -0,0 +1,80 @@
+      </button>
+      <div id="date" class="navbar-brand">
+        <% if @count -%>
+        <span id="badge_count" class="number_badge"><%= @count %></span>
```
After #1924, port the badge styles out of legacy and into `.bootstrap`. That will help us keep straight what's actually needed under the new design.

dnrce commented on app/assets/stylesheets/tracks.css.scss:
```diff
@@ -244,74 +244,8 @@ a.show_successors:hover, a.link_to_successors:hover {background-image: image-url
-    z-index:501;
-}
-
 #date {
```
After #1924, can `#date` be removed entirely?

dnrce commented on app/views/shared/_navbar.html.erb:
```diff
@@ -0,0 +1,80 @@
+        <span class="icon-bar"></span>
+        <span class="icon-bar"></span>
+      </button>
+      <div id="date" class="navbar-brand">
```
Remove after #1924?

matjack1 commented:
A huge thanks @dnrce. I'll work a bit more on this soon.

Regarding the font-awesome, with your legacy-ui change we could remove this namespace: https://github.com/TracksApp/tracks/blob/master/app/assets/stylesheets/application.css.scss#L18 and have the default glyphicons of Bootstrap working. So I could use that.

Instead regarding the theme, if the inverted style is good enough would be super easy to style everything. I must admit that would be nice to keep a bit of the current Tracks "feel". Even if it's not much, we could keep the logout link in red. But anyway this is just a small detail.

I'll let you know soon!

matjack1 commented:
@dnrce I've updated the PR with your suggestions. Now the PR is rebased over #1924, so it contains also your commit.

I've only left ```#menu_view_link``` and ```#menu_view_toggle_contexts``` as they are used ```tracks_step_helper.rb```.

I've also added a font-size adjustment for bootstrap, since I really think we should increase the font size. 14px feels already small to me, but less than that it's really bad for readability.

I'm happy to work more on this PR if anybody has any suggestion.

dnrce commented:
Bootstrap's default font size is larger than Tracks' existing font, but at the moment you're fighting against the `font-size: 80%` rule the legacy style puts on `body`, while at the same time Bootstrap's styles are _not_ being applied to `body`.

I'll see if I can tweak my CSS namespacing PR to improve this.

matjack1 commented:
Thanks @dnrce it's better now! I've rebased over your updated PR.
I'm still trying to figure out why the tests are failing. I still need to spend some time on that, but I wanted to let you know that your #1924 works well for me.

mattr- commented:
For the lazy among us, can we get screenshots of the before and after? Once of the goals that I had with the initial work of injecting bootstrap into the project to do less sweeping UI overhauls and more subtle changes and UI cleanups before really getting into massive amounts of UI changes.

matjack1 commented:
Sure @mattr- :)

There you go three screenshots of the navigation.

I tried working on a smaller thing, but even in the simplest page that I could find, the REST API doc, you still have the navigation that prevents us from going responsive, so I thought to tackle this problem first, trying to keep it as simple as I could. Anyway as you can see it's just a basic implementation of the Bootstrap navigation, basically just this: http://getbootstrap.com/examples/navbar/

I still need to work on the tests. If any of you have any idea is more then welcome :)

![development _tracks_list_tasks_-_2015-08-31_22 02 29](https://cloud.githubusercontent.com/assets/65402/9590411/507e48da-502c-11e5-8bdf-1395c775af21.png)
![development _tracks_list_tasks_-_2015-08-31_22 02 39](https://cloud.githubusercontent.com/assets/65402/9590410/507dd6c0-502c-11e5-8b70-1299ef56d88a.png)
![development _tracks_list_tasks_-_2015-08-31_22 03 40](https://cloud.githubusercontent.com/assets/65402/9590409/507ce760-502c-11e5-90d6-7fd943bd8564.png)

dnrce commented:
@matjack1, can you please rebase this again?

matjack1 commented:
@dnrce I've rebased, but I'm stuck with the failing tests. I've tried for a while now, but no joy with Superfish. Do you have any suggestions? I've tried with both Webkit and Selenium, same result :(

matjack1 commented:
Hello guys, I'm back and I'm really stuck. I can't replicate the errors with Selenium on my local machine. Some errors are similar, but not the same, using Phantom, and I can't figure out what is wrong. If you could try to run the tests yourself just because I'm curious to see if we have the same behaviour. Any idea or suggestion is really welcome.

mattr- commented:
I'll take a look at the tests on this one soon.

I had a thought about the navbar though. Could reduce the padding between the items to reduce the amount of horizontal space it takes up on for the desktop version? (you don't have to update the PR, just curious to see what it looks like comparing the old, this version, and this version with the reduced padding).

matjack1 commented:
Hey @mattr- would be awesome if you could have a look at the tests, because otherwise I don't know how to go ahead :(

I've done two tests, one with 7px padding:
![development _tracks_list_tasks_-_2015-12-01_18 02 59](https://cloud.githubusercontent.com/assets/65402/11507772/d41a547e-9855-11e5-8841-af89de68a74a.png)

and one with 10px:
![development _tracks_list_tasks_-_2015-12-01_18 03 08](https://cloud.githubusercontent.com/assets/65402/11507776/d99cc076-9855-11e5-897e-086425dbb9d8.png)

I wouldn't go under 7px otherwise they get too close.

I think that being a bit spaced apart makes them easier to read and to click, but I don't mind making them a bit closer to each other.

mattr- commented:
After seeing it again, I'm ok with the padding as is.

For the tests, I could only recreate a couple of the test failures and i'll work on fixing those soon.

mattr- commented:
I fixed up the one set of failing tests that I could reproduce and then merged this PR into master. Thanks for the contribution @matjack1!!

matjack1 commented:
Awesome!

C-Otto commented:
See #1947
